### ISSUE TITLE
Test loading NetFix at the wrong base address

### ISSUE THREAD

**Issue by Brett208** [OutpostUniverse/NetFixClient#25]
We updated the project file to include the linker property /FIXED (fixed base address: yes). 

This should stop the DLL from loading if it cannot use the specified base address of 0x14000000. 

There is code in NetFixClient to warn or abort loading the DLL if the wrong address is used. I suspect we may want to remove the FIXED property to allow for the better error message. Putting this issue in as a placeholder until we get reliably compiling to go back and test and determine if /FIXED should be removed.

https://github.com/OutpostUniverse/NetFixClient/blob/73839aa82e10515872987c93b1f740fd2daa6069/Main.cpp#L41-L46

**Comment by Brett208** [comment]
By commenting out 

//OPUNetGameProtocol opuNetGameProtocol;

and

// Set a new multiplayer protocol type
//protocolList[protocolIndex].netGameProtocol = &opuNetGameProtocol;

I was able to get NetFix to run. However, I received the message saying DLL loaded to bad address.

See attached screenshot. desiredLoadAddress is listed as 0x11000000 instead of 0x14000000. Guess I'm not sure if the project setting of 0x14000000 is wrong or if desiredLoadAddress pulling from op2Internal is wrong (or maybe the wrong thing is being checked on line 42).

<img width="768" alt="DllLoadAddress" src="https://user-images.githubusercontent.com/15183337/54094115-d7f5fa80-4374-11e9-9e21-abf5d081a53f.png">

Thanks, 
Brett

**Comment by DanRStevens** [comment]
NetFixClient uses a modified OP2Internal, which has the desired load address set to the NetFixClient settings. This change is not committed to the OP2Internal repository, and shouldn't be, as any levels that use OP2Internal would then conflict with NetFix. It's messy, but then so is the method being used to link to internal Outpost2.exe stuff.

As planned changes to OP2Internal should eliminate the need for both the `/Fixed` option, and the startup warning message, I think we should punt on this issue for now. Though I wouldn't be opposed to removing `/Fixed` if you think the startup error message is better.

In terms of compiling NetFixClient, be aware it will need that small local modification to OP2Internal.

**Comment by Brett208** [comment]
How do you modify OP2Internal to set the load address? I see the property desiredLoadAddress in the .asm file and in OP2Internal.h file, but it isn't set in either. There isn't a project level property to set fixed base address for a static library project as it is set by NetFixClient's properties.

This might be the last setting to get it actually working with op2ext.

-Brett

**Comment by DanRStevens** [comment]
Sorry, I should have provided details in my last post. It's set in the .asm file on line 15:
```asm
LoadAddr EQU 0x11000000
```

For NetFixClient, it should be modified to:
```asm
LoadAddr EQU 0x14000000
```

**Comment by Brett208** [comment]
Hey, making this change allows NetFixClient to run properly with Outpost 2. I can even load both NetFixClient and op2ext into the debugger at the same time and it doesn't complain. 

I haven't tested actually joining two games, but this is a good step in the right direction. :)

While the code to test the base address works, it cannot execute. If the dll is loaded to the wrong base address, initialization of NetFixClient will throw read access violations before it has a chance to check the base address.

Based on @DanRStevens' comments, fixing this issue can wait since it may be irrelevant soon. Guess we can leave it open until then as a reminder.

**Comment by DanRStevens** [comment]
The check should run before other code (other than static initializers and `DllMain`, which shouldn't be doing anything that touches Outpost2.exe's address space). The read access violation sounds like a legitimate bug.

We should probably investigate, as a low priority item. Even though we may change the project to eliminate the need for the check, this may be the symptom of a deeper bug which will simply be masked by future changes. We don't need a solution at this point, but a known root cause would be nice.

I took a quick look at DllMain and static initializers, but didn't see anything obvious.

Agreed that we should leave this open meanwhile as a reminder.

**Comment by Brett208** [comment]
I think this is caused by the following:
 - On module startup, `OPUNetGameProtocol opuNetGameProtocol` is initialized before InitMod is called.
 - OPUNetGameProtocol is a child of NetGameProtocol (from OP2Internal)
 - The module cannot find the virtual functions that OPUNetGameProtocol overrides because of the load address being wrong, causing a fatal error.

By placing the instance of `opuNetGameProtocol` in a raw pointer, and then only initializing it after the load address check, everything works. I briefly tried using a unique pointer, but couldn't figure out how to translate that into a raw pointer at line:

https://github.com/OutpostUniverse/NetFixClient/blob/779e845b0485e6b2c397a05d9e82aae062bacdd4/Main.cpp#L65

**Comment by DanRStevens** [comment]
Very interesting. That's a pretty good test you ran.

I checked the construction of `OPUNetGameProtocol` last time, and I just checked it again now. I'm not seeing the exact problem. That class, and parent class don't have constructors, so static initialization should be minimal. There doesn't seem to be any calls to those objects until after the check, so it's not obvious what's triggering the problem.

I do know the virtual function table pointer will need to be set when the object is constructed, but that should be a write into the DLL's module address space, not a read from Outpost2.exe's module address space.

I also looked at contained objects. The `NetGameProtocol` object has no member variables, only pure virtual functions. The `OPUNetGameProtocol` object contains an `OPUNetGameSelectWnd` object. That will be constructed. The `OPUNetGameSelectWnd` has a simple constructor that sets member variables, but doesn't read anything, particularly nothing from Outpost2.exe's module address space. No member objects in `OPUNetGameSelectWnd` should be problematic. None of them have any kind of constructor.

It's probably going to turn out to be something strangely obvious, just hidden from view. I'll look more into this later.

Thank you for the test results.

----

Edit: I think I see it.

`OPUNetGameSelectWnd` derives from `IDlgWnd`, which has a constructor. The constructor is not defined in NetFixClient, but rather imported by OP2Internal. That means construction can make a call into Outpost2.exe's module address space. Additionally, `IDlgWnd` derives from `IWnd`, which also has a constructor. Again, this is brought in by OP2Internal from Outpost2.exe's module address space.

Ok, that gives me some idea how to fix it. Likely a tomorrow problem though.
